File: rust/src/code/lru_cache.rs
use std::{
    cell::RefCell,
    collections::HashMap,
    rc::{Rc, Weak},
};

struct ListNode {
    prev: Option<Weak<RefCell<ListNode>>>,
    next: Option<Rc<RefCell<ListNode>>>,
    key: i32,
    value: i32,
}

impl ListNode {
    pub fn new(key: i32, value: i32) -> Self {
        ListNode {
            prev: None,
            next: None,
            key,
            value,
        }
    }

    fn pop(node: Rc<RefCell<ListNode>>) -> Rc<RefCell<ListNode>> {
        match (node.borrow().next.is_some(), node.borrow().prev.is_some()) {
            (true, true) => {
                let next = Rc::clone(node.borrow().next.as_ref().unwrap());
                let prev = Rc::clone(&node.borrow().prev.as_ref().unwrap().upgrade().unwrap());
                next.borrow_mut().prev = Some(Rc::downgrade(&prev));
                prev.borrow_mut().next = Some(Rc::clone(&next));
            }
            (true, false) => {
                node.borrow_mut().prev = None;
            }
            (false, true) => {
                node.borrow_mut().next = None;
            }
            _ => (),
        }

        node
    }
}

struct LRUCache {
    map: HashMap<i32, Option<Rc<RefCell<ListNode>>>>,
    head: Rc<RefCell<ListNode>>,
    tail: Rc<RefCell<ListNode>>,
    capacity: usize,
}

/**
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl LRUCache {
    fn new(capacity: i32) -> Self {
        // head and tail are markers and thus would not change, like a dummy head.
        let head = Rc::new(RefCell::new(ListNode::new(0, 0)));
        let tail = Rc::new(RefCell::new(ListNode::new(0, 0)));
        head.borrow_mut().next = Some(Rc::clone(&tail));
        tail.borrow_mut().prev = Some(Rc::downgrade(&head));

        LRUCache {
            map: HashMap::new(),
            head,
            tail,
            capacity: capacity as usize,
        }
    }

    fn get(&mut self, key: i32) -> i32 {
        match self.map.get(&key) {
            Some(value) => {
                if let Some(n) = value {
                    let node = ListNode::pop(Rc::clone(n));
                    let result = node.borrow().value;
                    Self::insert_after_head(self, node);

                    return result;
                } else {
                    -1
                }
            }
            None => -1,
        }
    }

    fn put(&mut self, key: i32, value: i32) {
        if let Some(v) = self.map.get(&key) {
            let node = v.as_ref().unwrap();
            let node = ListNode::pop(Rc::clone(node));
            node.borrow_mut().value = value;
            Self::insert_after_head(self, node);
        } else {
            if self.map.len() >= self.capacity {
                let del_node = self.tail.borrow().prev.as_ref().unwrap().upgrade().unwrap();
                let del_node = ListNode::pop(del_node);
                self.map.remove(&del_node.borrow().key);
            }

            let new_node = Rc::new(RefCell::new(ListNode::new(key, value)));
            self.map.insert(key, Some(Rc::clone(&new_node)));
            Self::insert_after_head(self, new_node);
        }
    }

    fn insert_after_head(&self, node: Rc<RefCell<ListNode>>) {
        let next = self.head.borrow_mut().next.take();
        next.as_ref().unwrap().borrow_mut().prev = Some(Rc::downgrade(&node));
        node.borrow_mut().next = next;
        node.borrow_mut().prev = Some(Rc::downgrade(&self.head));

        self.head.borrow_mut().next = Some(node);
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * let obj = LRUCache::new(capacity);
 * let ret_1: i32 = obj.get(key);
 * obj.put(key, value);
 */
#[cfg(test)]
mod tests {
    use super::LRUCache;

    #[test]
    fn case_1() {
        // arrange
        let mut cache = LRUCache::new(2);

        // act
        cache.put(1, 1); // cache is {1=1}
        cache.put(2, 2); // cache is {1=1, 2=2}
        let actual1 = cache.get(1); // return 1
        cache.put(3, 3); // LRU key was 2, evicts key 2, cache is {1=1, 3=3}
        let actual2 = cache.get(2); // returns -1 (not found)
        cache.put(4, 4); // LRU key was 1, evicts key 1, cache is {4=4, 3=3}
        let actual3 = cache.get(1); // return -1 (not found)
        let actual4 = cache.get(3); // return 3
        let actual5 = cache.get(4); // return 4

        // assert
        assert_eq!(actual1, 1);
        assert_eq!(actual2, -1);
        assert_eq!(actual3, -1);
        assert_eq!(actual4, 3);
        assert_eq!(actual5, 4);
    }
}

File: python/lru_cache.py
from typing import Optional


class ListNode:
    def __init__(
        self,
        key: int = 0,
        value: int = 0,
        next_node: Optional["ListNode"] = None,
        prev: Optional["ListNode"] = None,
    ):
        self.key: int = key
        self.value: int = value
        self.next: Optional[ListNode] = next_node
        self.prev: Optional[ListNode] = prev


class LRUCache:

    def __init__(self, capacity: int):
        self._capacity: int = capacity
        self._head: Optional[ListNode] = ListNode()
        self._tail: Optional[ListNode] = ListNode()
        self._head.next = self._tail
        self._tail.prev = self._head
        self._node_map: dict[int, ListNode] = {}

    def _insert_after_head(self, node: ListNode) -> None:
        """time complexity: O(1)"""
        next_node: ListNode = self._head.next
        next_node.prev = node
        self._head.next = node
        node.prev = self._head
        node.next = next_node

    def _pop_node(self, node: ListNode) -> Optional[ListNode]:
        """time complexity: O(1)"""
        next_node: ListNode = node.next
        prev_node: ListNode = node.prev
        next_node.prev = prev_node
        prev_node.next = next_node
        return node

    def get(self, key: int) -> int:
        """time complexity: O(1)"""
        if key not in self._node_map:
            return -1
        node: ListNode = self._node_map[key]
        iso_node: Optional[ListNode] = self._pop_node(node)
        self._insert_after_head(iso_node)
        return iso_node.value

    def put(self, key: int, value: int) -> None:
        """time complexity: O(1)"""
        if key not in self._node_map:
            if len(self._node_map) >= self._capacity:
                least_node: ListNode = self._tail.prev
                iso_least_node: ListNode = self._pop_node(least_node)
                del self._node_map[iso_least_node.key]
            new_node: ListNode = ListNode(key, value)
            self._node_map[key] = new_node
            self._insert_after_head(new_node)
            return
        node: ListNode = self._node_map[key]
        iso_node: ListNode = self._pop_node(node)
        iso_node.value = value
        self._insert_after_head(iso_node)


def test_lru_case1():
    # Your LRUCache object will be instantiated and called as such:
    # obj = LRUCache(capacity)
    # param_1 = obj.get(key)
    # obj.put(key,value)
    lruCache = LRUCache(2)
    lruCache.put(1, 1)  # cache is {1=1}
    lruCache.put(2, 2)  # cache is {1=1, 2=2}
    actual1 = lruCache.get(1)  # return 1
    lruCache.put(3, 3)  # LRU key was 2, evicts key 2, cache is {1=1, 3=3}
    actual2 = lruCache.get(2)  # returns -1 (not found)
    lruCache.put(4, 4)  # LRU key was 1, evicts key 1, cache is {4=4, 3=3}
    actual3 = lruCache.get(1)  # return -1 (not found)
    actual4 = lruCache.get(3)  # return 3
    actual5 = lruCache.get(4)  # return 4

    assert actual1 == 1
    assert actual2 == -1
    assert actual3 == -1
    assert actual4 == 3
    assert actual5 == 4
